addons/binding/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/discovery/NanoleafMDNSDiscoveryParticipant.java

```diff
@@ -0,0 +1,108 @@
+package org.openhab.binding.nanoleaf.internal.discovery;
+
+import static org.openhab.binding.nanoleaf.internal.NanoleafBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jmdns.ServiceInfo;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.mdns.MDNSDiscoveryParticipant;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.nanoleaf.internal.NanoleafBindingConstants;
+import org.openhab.binding.nanoleaf.internal.NanoleafHandlerFactory;
+import org.openhab.binding.nanoleaf.internal.OpenAPIUtils;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link NanoleafMDNSDiscoveryParticipant} is responsible for discovering new Nanoleaf controllers (bridges).
+ *
+ * @author Martin Raepple - Initial contribution
+ */
+
+@Component(immediate = true, configurationPid = "discovery.nanoleaf")
+@NonNullByDefault
+public class NanoleafMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
+
+    // see http://forum.nanoleaf.me/docs/openapi#_gf9l5guxt8r0
+    private static final String SERVICE_TYPE = "_nanoleafapi._tcp.local.";
+
+    private final Logger logger = LoggerFactory.getLogger(NanoleafMDNSDiscoveryParticipant.class);
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return NanoleafHandlerFactory.SUPPORTED_THING_TYPES_UIDS;
+    }
+
+    @Override
+    public String getServiceType() {
+        return SERVICE_TYPE;
+    }
+
+    @Override
+    public @Nullable DiscoveryResult createResult(ServiceInfo service) {
+        final ThingUID uid = getThingUID(service);
+        if (uid == null) {
+            return null;
+        }
+        final Map<String, Object> properties = new HashMap<>(2);
+        String host = service.getHostAddresses()[0];
+        properties.put(CONFIG_ADDRESS, host);
+        int port = service.getPort();
+        properties.put(CONFIG_PORT, port);
+        String firmwareVersion = service.getPropertyString("srcvers");
+
+        logger.debug("Nanoleaf controller with FW version {} found at {} {}", firmwareVersion, host, port);
+
+        if (OpenAPIUtils.checkRequiredFirmware(firmwareVersion)) {
+            final DiscoveryResult result = DiscoveryResultBuilder.create(uid).withThingType(getThingType(service))
+                    .withProperties(properties).withLabel("Nanoleaf Controller").build();
+            logger.debug("Nanoleaf controller added to inbox: {} at {}", uid.getId(), host);
+            return result;
+        } else {
+            logger.error("Nanoleaf controller firmware is too old: {}. Must be equal or higher than {}",
```

Comment:
According to the coding guidelines, error() should only be used if the whole system is in an unsuable state. I remember the convention being that a binding not working correctly does not qualify for that. Maybe just leave the brige in OFFLINE, with the appropiate state message telling the reason to the user?